---
layout: introduction_slides
logo: "GTN"

title: "ATAC-Seq data analysis"
type: "introduction"
contributors:
  - heylf
---

# What is ATAC sequencing?

---

### Where my data comes from?

![](../images/atac-seq/atac.jpg)

<small>
[*Buenrostro et al. 2013 Nat Methods*](https://doi.org/10.1002/0471142727.mb2129s109)
</small>

???

Usage of hyperactive Tn5 transposase to insert sequencing adapters into open chromatin regions.

It requires sequence-specific excision of a locus containing 19 base-pair inverted repeats.

After adapter attachment the DNA is sheared by the Transposase itself.

You have reads without nucleosomes (histones) = open chromatin regions, and reads with these complexes which will be filtered out because of their length.

---

### Characteristics of ATAC-Seq

- Typically biological replicates (at least 2).
- Control = genomic DNA from the sample that, instead of transposase treatment, is fragmented (e.g. by sonication), has adapters ligated, and is sequenced along with the ATAC sample.
- Paired-end sequencing, because you need both ends to idenfiy a true open chromatin region (both adapters are required).

---

## How to analyze ATAC-Seq data?

---

### Check the Insert Size

![](../images/atac-seq/Screenshot_sizeDistribution_Good.png)

???

Typical insert size of 150-200 bp and then a multiple of the number of nucelosomes.

The first peak (50bp) corresponds to where the Tn5 transposase inserted into nucleosome-free regions. The second peak (a bit less than 200bp) corresponds to where Tn5 inserted around a single nucleosome. The third one (around 400bp) is where Tn5 inserted around two adjacent nucleosomes and the fourth one (around 600bp) is where Tn5 inserted around three adjacent nucleosomes.

---

## Do not worry about a nucleotide bias

![](../images/atac-seq/per_base_sequence_content.png)

???

Your experiment might have a nucelotide bias because of transposase treatment depending on the enzyme ([Brian Green et. al. 2012](https://doi.org/10.1186/1759-8753-3-3))

---

## Filtering Reads

- Remove reads mapping to mitochondrial DNA
- Remove PCR duplicates
- Uniquely mapped reads with end-to-end alignment

???

ATAC-seq datasets usually contain a large percentage of reads that are derived from mitochondrial DNA. Since there are no ATAC-seq peaks of interest in the mitochondrial genome you can discard those reads.

End-to-End Alignement is important because you are intersted on the exact binding regions.

---

## Peak Calling

![](../images/atac-seq/schemeWithLegend.jpg)

- You need a peak caller taking into consideration that the adapters are separated by 9bp.
- Test Genrich and MACS2 both of them might produce different results based on the read coverage.

???

When Tn5 cuts an accessible chromatin locus it inserts adapters separated by 9bp [Kia et al. 2017](https://doi.org/10.1186/s12896-016-0326-1).
This means that to have the read start site reflect the centre of where Tn5 bound, the reads on the positive strand should be shifted 4 bp to the right and reads on the negative strands should be shifted 5 bp to the left as in [Buenrostro et al. 2013](https://doi.org/10.1002/0471142727.mb2129s109). Genrich can apply these shifts when ATAC-seq mode is selected.

---

## Overview

![](../images/atac-seq/ATACWF.svg)
