---
layout: tutorial_slides
logo: "GTN"
video: true

title: "Introduction to CRISPR screen analysis"
questions:
  - What is CRISPR?
  - What is a CRISPR screen?
  - How is a sgRNA library created?
  - What is the difference between a negative and positive screen?
  - How to analyse CRISPR screen data?
objectives:
  - Describe what CRISPR screen data is
  - Outline how CRISPR screen data is generated and analysed
requirements:
time_estimation: "5m"
contributors:
  - mblue9
  - sylviamahara

---


### What is CRISPR?


.image-100[![](../../images/crispr-screen/what_is_crispr.png)]

.center[Overview of CRISPR knockout method]


???

- CRISPR stands for Clustered Regularly Interspaced Short Palindromic Repeats.
- It's a bacterial immune system that has been modified for genome editing.
- It consists of 2 components - a guide RNA and a non-specific CRISPR-associated endonuclease called Cas9.
- The guide RNA is a short synthetic RNA composed of a scaffold sequence necessary for Cas9-binding and ~20 nucleotide spacer sequence that binds to the genomic target.
- Cas9 induces a double-stranded break within the target DNA.
- This results in in-frame amino acid deletions, insertions or frameshift mutations leading to premature stop codons within the targeted gene.
- With CRISPR knockout methods, ideally the end result is a loss-of-function mutation within the targeted gene.
- There are also CRISPR inhibition and activation methods but in this tutorial we focus on knockout.

---

### What is a CRISPR screen?

.image-100[![](../../images/crispr-screen/what_is_a_crispr_screen.png)]

???

- The ease of generating guide RNAs makes CRISPR one of the most scalable genome editing technologies.
- It can be used for genome-wide screens, enabling systematic targeting of 1000s of genes, with one gene targeted per cell.
- With these screens we can identify genes driving phenotypes, such as cell survival, drug resistance or sensitivity.
- Pooled or arrayed screens can be performed. In this tutorial we analyse data from a pooled screen.

---

- How is an sgRNA library created?

.image-50[![](../../images/crispr-screen/how_sgRNA_library_created.png)]

.pull-left[

* Various sgRNA libraries available e.g Brunello, Gecko, Yusa and others

* In this tutorial we use Brunello library (77,441 sgRNAs: ~4 per human gene + 1000 non-targeting controls)

* Cas9 expressing cells are transfected with sgRNA library

]

---

### What is the difference between a negative and positive screen?

.image-100[![](../../images/crispr-screen/pos_neg_screen.png)]

.right[Source: https://www.addgene.org/guides/pooled-libraries/]

???

- CRISPR positive or negative selection screens can be performed.
- With a positive selection screen, most cells die after the treatment (selection) and we are interested in identifying genes whose sgRNAs increase and dominate, indicating loss of those genes helps cells survive that treatment.
- With a negative selection screen, most cells survive after the treatment. In that case, we are interested in identifying genes whose guide RNAs decrease (drop out) compared to a control (e.g. vehicle), indicating those genes are needed for the cells to survive with that treatment.
- In this tutorial we analyse data from a negative screen.


---

## A workflow for CRISPR screen analysis

![](../../images/crispr-screen/crispr_workflow.png)


???

- This is an overview of the workflow used in the tutorial.
- We prepare the reads checking quality and trimming adapters.
- We use MAGeCK to analyse, followed by visualising results using volcano plots and identifying pathways.

