---
layout: tutorial_slides
logo: "GTN"

title: "High Performance Computing for Large-Scale Genome Comparison"
zenodo_link: "https://doi.org/10.5281/zenodo.61771"
questions:
  - What is sequence comparison and why does it matter?
  - What are the challenges and how can we use HPC to overcome them?
  - How can we use Galaxy for large scale sequence comparison?
objectives:
  - Learn the basics of genome comparison
  - Run sequence comparisons in Galaxy
  - Using GECKO for sequence comparison
  - Using GECKO-MGV for interactive visualization of sequence comparisons
  - Using CHROMEISTER for comparing large sequences such as plants and chromosomes
requirements:
time_estimation: "2h"
---

### What is sequence comparison?

- Finding string matches between text paragraphs
- Comparing biological sequences can be seen as an instance of common string matching problem but with some particularities

 - Restricted alphabets: e.g. DNA = \{A, C, G, T\}
 - It adds an additional dimension
  - Instead of looking for one pattern, we look for all possible patterns!
---

### Why is it important?

- Sequence comparison is a core problem of bioinformatics that allows us to study:
 - Evolution (Where do we come from? And other species?)
 - Structure (What are we made of?)
 - Functionality (How do DNA changes affect us?)

- Widely used in several fields!
 - Whole genome alignment, gene prediction, phylogenomics, pharmacogenetics, metagenomics, etc.

---

### What are the challenges?

- A lot of progress has taken place since the beginning (around 1960s)
- So what is the problem?

 - Optimal sequence comparison has complexity O(n<sup>2</sup>)
 - **New technology produces huge amounts of data**
 - Good mathematical models exist for comparing small sequences, but not for large ones
 - Assessing statistical significance of findings depends on several variables

---

### Why is sequence comparison hard?

- The complexity of the problem grows as the length of the sequences become larger
- Computing power doubles every two years according to Moore’s law but sequence databases grow purely exponential every year

- Large-sized genomics are more fun!


Sequence type    | Approx. length
---        | ---          
Gene			|  ~1,000 to 10,000 bp
Bacteria genome		|  ~10<sup>6</sup> bp
Mammalian chr.		|  ~10<sup>8</sup> bp
Mammalian genome	|  ~2 * 10<sup>9</sup> bp
Plant genome		|  ~3 * 10<sup>9</sup> bp
Large plant genomes	|  ~2 * 10<sup>10</sup> bp
Amoeba genome		|  ~6 * 10<sup>11</sup> bp


---
### Exponential growth of the Sequence Read Archive


![](../../images/hpc-for-lsgc/SRA.png)

- Other databases have experienced similar growth too

---

### Methods for sequence comparison

- Dot-matrix comparison (no alignments)
 - Comparing all letters with each other (extremely slow/noisy)

- Aligning methods
 - Exhaustive dynamic programming algorithms (Slow but optimal)
  - Local alignment
  - Global alignment
 - Seed-based heuristics
  - Words of fixed size are used to start alignments (Fast but less sensitivity) 

---

### Dot-matrix methods


- Takes each letter in both sequences and compare it to all the rest
 - Takes quadratic time and space!
- Does not produce alignments, but rather just represents a comparison
- The x and y axis represent the sequences being compared

![](../../images/hpc-for-lsgc/Dot01.png)

---

### Comparing genomes with dotplots

![](../../images/hpc-for-lsgc/Dot02.png)

---

###  wat

![](../../images/from_exp_to_data.png)

Quality control = First step of the bioinformatics analyses

???

- But no matter what technique used, same file format
- Always should go through QC

Segue: So let's look at how that data is stored

---

### Sequences: FASTA

```
>Identifier1 (comment)
XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
>Identifier2 (comment)
XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
XX
```

???

- familiar if you've seen genomic data before
- Starts with greater-than sign
- Then an identifier with no spaces
- Everything after a space is a comment
- Newline
- Everything up to next greater-than is sequence data (wrapping unnecessary)

Segue: But this is just sequence, and we have data from a sequencer, which includes quality

---

### Sequences: FASTQ

```
@Identifier1 (comment)
XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
+
QQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQ
@Identifier2 (comment)
XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
+
QQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQ
```

???

- Four lines
- @ + identifier on first line, just like fasta
- sequence
- +
- quality score characters

Segue: so what do the quality chars mean?

---

### Quality score

Measure of the quality of the identification of the nucleobases <br>generated by automated DNA sequencing

<small>

Phred Quality Score | Probability of incorrect base call | Base call accuracy
--- | --- | ---
10 | 1 in 10 | 90%
20 | 1 in 100 | 99%
30 | 1 in 1000 | 99.9%
40 | 1 in 10,000 | 99.99%
50 | 1 in 100,000 | 99.999%
60 | 1 in 1,000,000 | 99.9999%

</small>

???

- Logarithmic scale

---

### Quality score

![](../../images/quality_score.png)

???

- Solexa uses slightly different measure

---

### Quality score encoding

![](../../images/quality_score_encoding.png)

???

- Earlier saw sequence + quality were same length of text
- Each score encoded as a single character
- Uses ASCII (common computer encoding, maps numbers to characters)
- Each technology has a different mapping
- Mostly see illumina 1.8 these days, others only if historical

---

## Identifying Potential Quality Issues

---

### FastQC

![](../../images/fastqc-report.png)

???

- Analyses your fastq files
- Has many modules
- They report on different aspects of the sequence

- Modules include:
  - Basic Statistics
  - Per base sequence quality
  - Per sequence quality scores
  - Per base sequence content
  - Per base GC content
  - Per sequence GC content
  - Per base N content
  - Sequence Length Distribution
  - Sequence Duplication Levels
  - Overrepresented sequences
  - Kmer Content

---

### Quality score: Per-base

![](../../images/per_base_sequence_quality_good.png)

<i class="fa fa-thumbs-up"></i> Good quality score

---


### Per-base Quality

![](../../images/per_base_sequence_quality_bad.png)

<i class="fa fa-thumbs-down"></i> Bad quality score

---

### Per-base Quality

![](../../images/per_base_sequence_quality_medium.png)

<i class="fa fa-thumbs-up"></i> <i class="fa fa-thumbs-down"></i> Intermediate quality score

---

### Per-sequence Quality

![](../../images/per_sequence_quality.png)

---

### Per-tile Quality

![](../../images/per_tile_quality.png)

???
In Illumina libraries, the original sequence identifier is retained. Encoded in these is the flowcell tile from which each read came.

There might be transient problems such as bubbles going through the flowcell, or more permanent problems such as smudges on the flowcell, or debris inside the flowcell lane.

This graph will only appear with Illumina libraries which retain their original sequence identifiers. The graph allows to check the quality scores from each tile across all bases, to see if there was a loss in quality associated with only one portion of the flowcell. The plot shows the deviation from the average quality for each tile. The colours are on a cold to hot scale, with cold colours being positions where the quality was at or below the average for that base in the run, and hot colours to indicate that a tile had worse quality reads than other tiles for that base. In the example below you can see that certain tiles show consistently poor quality. A good plot should be blue all over.

---

### Per-base Sequence Content

![](../../images/per_base_sequence_content.png)

???
The per-base sequence content highlights the proportion of each base in each position of a sequence for which each of the four DNA bases have been called. In a random library there would be little to no difference between the different bases of a sequence run. The relative amount of each base should reflect the overall amount of these bases, but in any case they should not be hugely imbalanced from one another. It is worth noting that some types of libraries will always produce biased sequence composition, normally at the start of the read. Libraries produced by priming with random hexamers (including nearly all RNA-Seq libraries) and those which were fragmented using transposases, inherit an intrinsic bias in the positions at which the reads start. This bias does not concern an absolute sequence, but instead provides an enrichment of a number of different K-mers at the 5’ end of the reads. Whilst this is a true technical bias, it isn’t something which can be corrected by trimming and in most cases doesn’t seem to adversely affect the downstream analysis. It will however produce a warning or error in this module.

There are a number of common scenarios for these issues:

- Over-represented sequences
- Biased fragmentation
- Biased composition libraries
- Aggressive adapter trimming

---

### Per-sequence GC content

![](../../images/per_sequence_GC_content.png)

???
The GC content distribution of most samples should follow a normal distribution. In some cases, a bi-modal distribution can be observed, especially for metagenomic data sets. An unusually shaped distribution could indicate a contaminated library or some other kinds of biased subset. A normal distribution which is shifted indicates some systematic bias which is independent of base position. Such a systematic bias creating a shifted normal distribution won’t be flagged as an error, since the tool cannot guess what the provided genome’s GC content should be.

Issues in the GC content distribution usually indicate a problem with the library. Sharp peaks on an otherwise smooth distribution are normally the result of a specific contaminant (adapter dimers for example), which may well be picked up by the over-represented sequences module. Broader peaks may represent contamination with a different species.

---

### Per-base N content

![](../../images/per_base_N_content.png)

???
Sequences can contain the ambiguous base N for positions that could not be identified as a particular base. A high number of Ns can be a sign for a low quality sequence or even dataset. If no quality scores are available, the sequence quality can be inferred from the percent of Ns found in a sequence or dataset.

If a sequencer is unable to make a base call with sufficient confidence then it will normally substitute an N rather than a conventional base call. It’s not unusual to see a very low proportion of Ns appearing in a sequence, especially nearer the end of a sequence. However, if this proportion rises above a few percent it suggests that the analysis pipeline was unable to interpret the data well enough to make valid base calls.

---

### Sequence length distribution

![](../../images/sequence_length_distribution.png)

???
Some high throughput sequencers generate sequence fragments of uniform length, while others can output reads of wildly varying lengths. The length distribution can be then used as quality measure. You would expect a normal distribution for the best result. However, most sequencing results show a slowly increasing and then a steep falling distribution.

FastQC generates a graph showing the distribution of fragment sizes in the file which was analysed. In many cases this will produce a simple graph showing a peak only at one size, but for variable length FASTQ files this will show the relative amounts of each different size of sequence fragment.

This module will raise a warning if all sequences are not the same length. This module will raise an error if any of the sequences have zero length.

---

### Duplicated sequences

![](../../images/sequence_duplication_levels.png)

???
This quality check module counts the degree of duplication for every sequence in the library, and creates a plot showing the relative number of sequences with different degrees of duplication:
- the blue line represents the full sequence set, showing how its duplication levels are distributed;
- the red line represents the de-duplicated sequences, plotting the proportions of deduplicated sequence sets which come from different duplication levels in the original data.

In genomic projects, sequence duplication is investigated. Duplicated sequences can arise when there are too few fragments present at any stage prior to sequencing.

This module issues a warning if non-unique sequences make up for more than 20% of the total sequences. An error is raised instead if non-unique sequences make up for more than 50% of the total.

---

### Tag sequences: Adapter contamination

![](../../images/adapter_content.png)

???
Tag sequences are artifacts at the ends of sequence reads such as multiplex identifiers, adapters, and primer sequences that were introduced during pre-amplification with primer-based methods. The base frequencies across the reads present an easy way to check for tag sequences. If the distribution seems uneven (high frequencies for certain bases over several positions), it could indicate some residual tag sequences. This doesn’t indicate a problem as such - just that the sequences will need to be adapter trimmed before proceeding with any downstream analysis.

To investigate tag or adapter content, FastQC generates a plot showing a cumulative percentage count of the proportion of the library which has seen each of the adapter sequences at each position. Once a sequence has been seen in a read it is counted as being present right through to the end of the read so the percentages you see will only increase as the read length goes on.

---

### Tag sequences: K-mer content

![](../../images/kmer_content.png)

???
Another way to find tag sequences is to look at the K-mer content, and find those which do not have even coverage through the length of your reads and could correspond to tag sequences.

K-mers with positionally biased enrichment are reported. The top 6 most biased K-mer are additionally plotted to show their distribution.

Over-represented K-mers will appear as sharp spikes at a single point in the sequence, deviating from what should be a progressive or broad enrichment.

---

## Improving the quality of sequences

- Filtering of sequences
    - with small mean quality score
    - too small
    - with too many N bases
    - based on their GC content
    - ...
- Cutting/Trimming sequences
    - from low quality score parts
    - tails
    - ...
